source/Host/common/Symbols.cpp

```diff
@@ -104,38 +103,41 @@ static bool LocateDSYMInVincinityOfExecutable(const ModuleSpec &module_spec,
           }
           return true;
         } else {
-          path[obj_file_path_length] = '\0';
-
-          char *last_dot = strrchr(path, '.');
-          while (last_dot != NULL && last_dot[0]) {
-            char *next_slash = strchr(last_dot, '/');
-            if (next_slash != NULL) {
-              *next_slash = '\0';
-              ::strncat(path, ".dSYM/Contents/Resources/DWARF/",
-                        sizeof(path) - strlen(path) - 1);
-              ::strncat(path, exec_fspec->GetFilename().AsCString(),
-                        sizeof(path) - strlen(path) - 1);
-              dsym_fspec.SetFile(path, false);
-              if (dsym_fspec.Exists() &&
-                  FileAtPathContainsArchAndUUID(
-                      dsym_fspec, module_spec.GetArchitecturePtr(),
-                      module_spec.GetUUIDPtr())) {
-                if (log) {
-                  log->Printf("dSYM with matching UUID & arch found at %s",
-                              path);
-                }
-                return true;
-              } else {
-                *last_dot = '\0';
-                char *prev_slash = strrchr(path, '/');
-                if (prev_slash != NULL)
-                  *prev_slash = '\0';
-                else
-                  break;
-              }
-            } else {
-              break;
+          // Get a clean copy of the executable path, without the final filename
+          FileSpec path_dir_fspec (exec_fspec->GetDirectory().AsCString(), true);
+          std::string path_dir_str = path_dir_fspec.GetPath();
+
+          // Add a ".dSYM" name to each directory component of the path, stripping
+          // off components.  e.g. we may have a binary like
+          // /S/L/F/Foundation.framework/Versions/A/Foundation
+          // and
+          // /S/L/F/Foundation.framework.dSYM
+          //
+          // so we'll need to start with /S/L/F/Foundation.framework/Versions/A,
+          // add the .dSYM part to the "A", and if that doesn't exist, strip off
+          // the "A" and try it again with "Versions", etc., until we find a dSYM
+          // bundle or we've stripped off enough path components that there's no
+          // need to continue.
+
+          for (int i = 0; i < 4; i++) {
```

Comment:
It looks to me like all these operations are supported by `FileSpec`, e.g. `CopyByRemovingLastPathComponent` for your `rfind`-based approach and `AppendPathComponent` instead of your `+=`.  In fact, it looks like the FindBundleBinaryInExecSearchPaths uses these APIs.  Can we avoid doing this as `std::string` processing?